diff --git a/cpp-tiny/README.mustache b/cpp-tiny/README.mustache
index dc417d1e..3350b54a 100644
--- a/cpp-tiny/README.mustache
+++ b/cpp-tiny/README.mustache
@@ -1,31 +1,9 @@
-# Documentation for OpenAPI Petstore
+# Documentation for Thy Street SDK
 This is a client generator for microcontrollers on the Espressif32 platform and the Arduino framework
-After the client have been generated, you have to change these following variables:
-- root.cert | Provide your service root certificate.
-- src/main.cpp | Change wifi name
-- src/main.cpp | Change wifi password
-- lib/service/AbstractService.h | Change to your url
 
-# Documentation for {{#openAPI}}{{#info}}{{title}} {{version}} Tiny client cpp (Arduino) {{/info}}{{/openAPI}}
+# Documentation for {{#openAPI}}{{#info}}{{title}} {{version}} cpp client (Arduino) {{/info}}{{/openAPI}}
 
-The project is structured like this:
-```
-samples/client/petstore/tiny/cpp/
-├── lib
-│   ├── Models
-│   ├── service
-│   └── TestFiles
-├── platformio.ini
-├── pre_compiling_bourne.py
-├── README.md
-├── root.cert
-├── src
-│   └── main.cpp
-└── test
-    └── RunTests.cpp
-```
-
-All URIs are relative to {{{scheme}}}://{{{host}}}{{{basePath}}}
+All URIs are relative to {{{basePath}}}
 {{#apiInfo}}{{#apis}}{{#operations}}
 ### {{classname}}
 |Method | HTTP request | Description|
diff --git a/cpp-tiny/main.mustache b/cpp-tiny/main.mustache
index 56170a91..694b8ace 100644
--- a/cpp-tiny/main.mustache
+++ b/cpp-tiny/main.mustache
@@ -1,4 +1,4 @@
-#include "PetApi.h"
+#include "DeviceApi.h"
 {{#isESP8266}}
 #include "ESP8266WiFi.h"
 {{/isESP8266}}
@@ -6,6 +6,8 @@
 const char* ssid = "your wifi name"; // TODO Change wifi name
 const char* password = "Your wifi password"; //TODO Change wifi password
 
+const char* apiKey = "Thy Street API key"; // TODO Change wifi name
+const char* apiSecret = "Thy Street API secret"; //TODO Change wifi password
 
 void setup(){
     //Initialize serial and wait for port to open:
@@ -30,11 +32,15 @@ void setup(){
     Serial.print("IP address set: ");
     Serial.println(WiFi.localIP());
 
-    Tiny::PetApi petapi;
-    auto resp = petapi.getPetById(10);
+    Tiny::DeviceApi api;
+    api.setAuthorization(apiKey, apiSecret);
+
+    Tiny::SetDeviceDetailsDto ddDto;
+    ddDto.setDeviceId("device-id");
+    ddDto.setTariff(pdTRUE);
+    auto resp = api.setDetails(ddDto);
     Serial.println(resp.code);
-    Tiny::Pet pet = resp.obj;
-    Serial.println(pet.toJson().dump().c_str());
+    Serial.println(resp.obj.c_str());
 }
 
 void loop(){}
diff --git a/cpp-tiny/root.cert.mustache b/cpp-tiny/root.cert.mustache
index bb78f18e..e334a431 100644
--- a/cpp-tiny/root.cert.mustache
+++ b/cpp-tiny/root.cert.mustache
@@ -1,53 +1,64 @@
-// TODO: Provide your service root certificate.
-//    Below is two examples of root certificates.
-
-// Let's encrypt root certificate
-
 "-----BEGIN CERTIFICATE-----\n" \
-"MIIDrzCCApegAwIBAgIQCDvgVpBCRrGhdWrJWZHHSjANBgkqhkiG9w0BAQUFADBh\n" \
-"MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3\n" \
-"d3cuZGlnaWNlcnQuY29tMSAwHgYDVQQDExdEaWdpQ2VydCBHbG9iYWwgUm9vdCBD\n" \
-"QTAeFw0wNjExMTAwMDAwMDBaFw0zMTExMTAwMDAwMDBaMGExCzAJBgNVBAYTAlVT\n" \
-"MRUwEwYDVQQKEwxEaWdpQ2VydCBJbmMxGTAXBgNVBAsTEHd3dy5kaWdpY2VydC5j\n" \
-"b20xIDAeBgNVBAMTF0RpZ2lDZXJ0IEdsb2JhbCBSb290IENBMIIBIjANBgkqhkiG\n" \
-"9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4jvhEXLeqKTTo1eqUKKPC3eQyaKl7hLOllsB\n" \
-"CSDMAZOnTjC3U/dDxGkAV53ijSLdhwZAAIEJzs4bg7/fzTtxRuLWZscFs3YnFo97\n" \
-"nh6Vfe63SKMI2tavegw5BmV/Sl0fvBf4q77uKNd0f3p4mVmFaG5cIzJLv07A6Fpt\n" \
-"43C/dxC//AH2hdmoRBBYMql1GNXRor5H4idq9Joz+EkIYIvUX7Q6hL+hqkpMfT7P\n" \
-"T19sdl6gSzeRntwi5m3OFBqOasv+zbMUZBfHWymeMr/y7vrTC0LUq7dBMtoM1O/4\n" \
-"gdW7jVg/tRvoSSiicNoxBN33shbyTApOB6jtSj1etX+jkMOvJwIDAQABo2MwYTAO\n" \
-"BgNVHQ8BAf8EBAMCAYYwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUA95QNVbR\n" \
-"TLtm8KPiGxvDl7I90VUwHwYDVR0jBBgwFoAUA95QNVbRTLtm8KPiGxvDl7I90VUw\n" \
-"DQYJKoZIhvcNAQEFBQADggEBAMucN6pIExIK+t1EnE9SsPTfrgT1eXkIoyQY/Esr\n" \
-"hMAtudXH/vTBH1jLuG2cenTnmCmrEbXjcKChzUyImZOMkXDiqw8cvpOp/2PV5Adg\n" \
-"06O/nVsJ8dWO41P0jmP6P6fbtGbfYmbW0W5BjfIttep3Sp+dWOIrWcBAI+0tKIJF\n" \
-"PnlUkiaY4IBIqDfv8NZ5YBberOgOzW6sRBc4L0na4UU+Krk2U886UAb3LujEV0ls\n" \
-"YSEY1QSteDwsOoBrp+uvFRTp2InBuThs4pFsiv9kuXclVzDAGySj4dzp30d8tbQk\n" \
-"CAUw7C29C79Fv1C5qfPrmAESrciIxpg0X40KPMbp1ZWVbd4=\n" \
-"-----END CERTIFICATE-----\n"
-
-
-// Amazon_Root_CA_1.pem
-
-"-----BEGIN CERTIFICATE-----\n" \
-"MIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF\n" \
-"ADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6\n" \
-"b24gUm9vdCBDQSAxMB4XDTE1MDUyNjAwMDAwMFoXDTM4MDExNzAwMDAwMFowOTEL\n" \
-"MAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEZMBcGA1UEAxMQQW1hem9uIFJv\n" \
-"b3QgQ0EgMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJ4gHHKeNXj\n" \
-"ca9HgFB0fW7Y14h29Jlo91ghYPl0hAEvrAIthtOgQ3pOsqTQNroBvo3bSMgHFzZM\n" \
-"9O6II8c+6zf1tRn4SWiw3te5djgdYZ6k/oI2peVKVuRF4fn9tBb6dNqcmzU5L/qw\n" \
-"IFAGbHrQgLKm+a/sRxmPUDgH3KKHOVj4utWp+UhnMJbulHheb4mjUcAwhmahRWa6\n" \
-"VOujw5H5SNz/0egwLX0tdHA114gk957EWW67c4cX8jJGKLhD+rcdqsq08p8kDi1L\n" \
-"93FcXmn/6pUCyziKrlA4b9v7LWIbxcceVOF34GfID5yHI9Y/QCB/IIDEgEw+OyQm\n" \
-"jgSubJrIqg0CAwEAAaNCMEAwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMC\n" \
-"AYYwHQYDVR0OBBYEFIQYzIU07LwMlJQuCFmcx7IQTgoIMA0GCSqGSIb3DQEBCwUA\n" \
-"A4IBAQCY8jdaQZChGsV2USggNiMOruYou6r4lK5IpDB/G/wkjUu0yKGX9rbxenDI\n" \
-"U5PMCCjjmCXPI6T53iHTfIUJrU6adTrCC2qJeHZERxhlbI1Bjjt/msv0tadQ1wUs\n" \
-"N+gDS63pYaACbvXy8MWy7Vu33PqUXHeeE6V/Uq2V8viTO96LXFvKWlJbYK8U90vv\n" \
-"o/ufQJVtMVT8QtPHRh8jrdkPSHCa2XV4cdFyQzR1bldZwgJcJmApzyMZFo6IQ6XU\n" \
-"5MsI+yMRQ+hDKXJioaldXgjUkK642M4UwtBV8ob2xJNDd2ZhwLnoQdeXeGADbkpy\n" \
-"rqXRfboQnoZsG4q5WTP468SQvvG5\n" \
+"MIIFjDCCA3SgAwIBAgINAgCOsgIzNmWLZM3bmzANBgkqhkiG9w0BAQsFADBHMQsw\n" \
+"CQYDVQQGEwJVUzEiMCAGA1UEChMZR29vZ2xlIFRydXN0IFNlcnZpY2VzIExMQzEU\n" \
+"MBIGA1UEAxMLR1RTIFJvb3QgUjEwHhcNMjAwODEzMDAwMDQyWhcNMjcwOTMwMDAw\n" \
+"MDQyWjBGMQswCQYDVQQGEwJVUzEiMCAGA1UEChMZR29vZ2xlIFRydXN0IFNlcnZp\n" \
+"Y2VzIExMQzETMBEGA1UEAxMKR1RTIENBIDFENDCCASIwDQYJKoZIhvcNAQEBBQAD\n" \
+"ggEPADCCAQoCggEBAKvAqqPCE27l0w9zC8dTPIE89bA+xTmDaG7y7VfQ4c+mOWhl\n" \
+"UebUQpK0yv2r678RJExK0HWDjeq+nLIHN1Em5j6rARZixmyRSjhIR0KOQPGBMUld\n" \
+"saztIIJ7O0g/82qj/vGDl//3t4tTqxiRhLQnTLXJdeB+2DhkdU6IIgx6wN7E5NcU\n" \
+"H3Rcsejcqj8p5Sj19vBm6i1FhqLGymhMFroWVUGO3xtIH91dsgy4eFKcfKVLWK3o\n" \
+"2190Q0Lm/SiKmLbRJ5Au4y1euFJm2JM9eB84Fkqa3ivrXWUeVtye0CQdKvsY2Fka\n" \
+"zvxtxvusLJzLWYHk55zcRAacDA2SeEtBbQfD1qsCAwEAAaOCAXYwggFyMA4GA1Ud\n" \
+"DwEB/wQEAwIBhjAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwEgYDVR0T\n" \
+"AQH/BAgwBgEB/wIBADAdBgNVHQ4EFgQUJeIYDrJXkZQq5dRdhpCD3lOzuJIwHwYD\n" \
+"VR0jBBgwFoAU5K8rJnEaK0gnhS9SZizv8IkTcT4waAYIKwYBBQUHAQEEXDBaMCYG\n" \
+"CCsGAQUFBzABhhpodHRwOi8vb2NzcC5wa2kuZ29vZy9ndHNyMTAwBggrBgEFBQcw\n" \
+"AoYkaHR0cDovL3BraS5nb29nL3JlcG8vY2VydHMvZ3RzcjEuZGVyMDQGA1UdHwQt\n" \
+"MCswKaAnoCWGI2h0dHA6Ly9jcmwucGtpLmdvb2cvZ3RzcjEvZ3RzcjEuY3JsME0G\n" \
+"A1UdIARGMEQwCAYGZ4EMAQIBMDgGCisGAQQB1nkCBQMwKjAoBggrBgEFBQcCARYc\n" \
+"aHR0cHM6Ly9wa2kuZ29vZy9yZXBvc2l0b3J5LzANBgkqhkiG9w0BAQsFAAOCAgEA\n" \
+"IVToy24jwXUr0rAPc924vuSVbKQuYw3nLflLfLh5AYWEeVl/Du18QAWUMdcJ6o/q\n" \
+"FZbhXkBH0PNcw97thaf2BeoDYY9Ck/b+UGluhx06zd4EBf7H9P84nnrwpR+4GBDZ\n" \
+"K+Xh3I0tqJy2rgOqNDflr5IMQ8ZTWA3yltakzSBKZ6XpF0PpqyCRvp/NCGv2KX2T\n" \
+"uPCJvscp1/m2pVTtyBjYPRQ+QuCQGAJKjtN7R5DFrfTqMWvYgVlpCJBkwlu7+7KY\n" \
+"3cTIfzE7cmALskMKNLuDz+RzCcsYTsVaU7Vp3xL60OYhqFkuAOOxDZ6pHOj9+OJm\n" \
+"YgPmOT4X3+7L51fXJyRH9KfLRP6nT31D5nmsGAOgZ26/8T9hsBW1uo9ju5fZLZXV\n" \
+"VS5H0HyIBMEKyGMIPhFWrlt/hFS28N1zaKI0ZBGD3gYgDLbiDT9fGXstpk+Fmc4o\n" \
+"lVlWPzXe81vdoEnFbr5M272HdgJWo+WhT9BYM0Ji+wdVmnRffXgloEoluTNcWzc4\n" \
+"1dFpgJu8fF3LG0gl2ibSYiCi9a6hvU0TppjJyIWXhkJTcMJlPrWx1VytEUGrX2l0\n" \
+"JDwRjW/656r0KVB02xHRKvm2ZKI03TglLIpmVCK3kBKkKNpBNkFt8rhafcCKOb9J\n" \
+"x/9tpNFlQTl7B39rJlJWkR17QnZqVptFePFORoZmFzM=\n" \
 "-----END CERTIFICATE-----\n" \
-
-
+"-----BEGIN CERTIFICATE-----\n" \
+"MIIFczCCBFugAwIBAgIQOmRmLLg9lPAQAAAAAA3UcjANBgkqhkiG9w0BAQsFADBG\n" \
+"MQswCQYDVQQGEwJVUzEiMCAGA1UEChMZR29vZ2xlIFRydXN0IFNlcnZpY2VzIExM\n" \
+"QzETMBEGA1UEAxMKR1RTIENBIDFENDAeFw0yMjAzMjAyMDEyMzRaFw0yMjA2MTgy\n" \
+"MDEyMzNaMCMxITAfBgNVBAMTGGNoZWZmeS1hcGkudGh5c3RyZWV0LmNvbTCCASIw\n" \
+"DQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAJmU7JLhUOQhY+Wa8Qd4S0fIsSAo\n" \
+"YZOvvbabbKprj7BPOPyy0zfRwNRFy9bhWUDekJG+P99CvCmPQabq9Iq6l0k830wd\n" \
+"uVCONgYPjcylB6RkWho27ODlVPvHpU5YJSdbE4lIZrYz/hOhjiPOHNkjtalVkz6s\n" \
+"PINGd3GXF2SQB9d9rHkVwALLYFsiqRWcSNFUm4ZiaVD+Oehd9ew4vF8t1yvvzz8o\n" \
+"FaLioRYE644Ym+z6VgYfScAYaxpK3idV9+XPUfhLDmMcd7Go7isdOFGKo3z1bWuB\n" \
+"hcXvZ/Ew07CFTdvAd31OmrEgAVhcilf3rH1nuUBUwk1WvRd4GBNNu59hmOECAwEA\n" \
+"AaOCAn4wggJ6MA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATAM\n" \
+"BgNVHRMBAf8EAjAAMB0GA1UdDgQWBBRdpBlHUNt5HdXCX/V2XruUlH2VZDAfBgNV\n" \
+"HSMEGDAWgBQl4hgOsleRlCrl1F2GkIPeU7O4kjB4BggrBgEFBQcBAQRsMGowNQYI\n" \
+"KwYBBQUHMAGGKWh0dHA6Ly9vY3NwLnBraS5nb29nL3MvZ3RzMWQ0L2NvVXQ2TWk2\n" \
+"eXpNMDEGCCsGAQUFBzAChiVodHRwOi8vcGtpLmdvb2cvcmVwby9jZXJ0cy9ndHMx\n" \
+"ZDQuZGVyMCMGA1UdEQQcMBqCGGNoZWZmeS1hcGkudGh5c3RyZWV0LmNvbTAhBgNV\n" \
+"HSAEGjAYMAgGBmeBDAECATAMBgorBgEEAdZ5AgUDMDwGA1UdHwQ1MDMwMaAvoC2G\n" \
+"K2h0dHA6Ly9jcmxzLnBraS5nb29nL2d0czFkNC9fcmFhWWJfbktQZy5jcmwwggED\n" \
+"BgorBgEEAdZ5AgQCBIH0BIHxAO8AdQBByMqx3yJGShDGoToJQodeTjGLGwPr60vH\n" \
+"aPCQYpYG9gAAAX+pORGAAAAEAwBGMEQCIC/QjvEUK1oLbQmhsAHigL/917OuwYYJ\n" \
+"M8F21oRlCb4WAiB2G3baQocS1xwJk6Hh2lisfUp9E/Rp+vW8wqQTy/YwCQB2AEal\n" \
+"Vet1+pEgMLWiiWn0830RLEF0vv1JuIWr8vxw/m1HAAABf6k5EY8AAAQDAEcwRQIh\n" \
+"AJq9MATR1wSOEoAzSZ8wSGlBW5ExOs0FLQYPdNZuTrp1AiAnHeTq7Poz40TwR7UX\n" \
+"9q7jKjBqAVkp4I+DSlSoaPxNzjANBgkqhkiG9w0BAQsFAAOCAQEAKKhxs3FWqGhC\n" \
+"ldK2hsjGV2Uw4sVKT6ARR0yFpP6TIVs0mhN7tv496dwYwTjT+zns9g1kViB3gyZ/\n" \
+"THJQCbJYjfU8aYxJue4ekecNQAI58B4+fJEbOBuaTFtzUjBFvxS17ledt9nq5NIc\n" \
+"21SS23TbRKVoSpSvRLNyGwlvfexA+kK9XItyjb1GDL5uTz0bO7D4pBMzxv88NyRt\n" \
+"0R+n9mJ9Fh4Gz9xKGFshSShaMQsvyJ+f/w/783+RU+LE/9+Oe/JEifM+/9Ezg2E3\n" \
+"BQp+uxP5DPYna3g5w+OdQJJJUI5MkeafofDzAW6mwxxi8Glha5VS9sMMIlsGVPd5\n" \
+"d+LlsYBmLA==\n" \
+"-----END CERTIFICATE-----\n"
\ No newline at end of file
diff --git a/cpp-tiny/service/Service.cpp.mustache b/cpp-tiny/service/Service.cpp.mustache
index d164c311..1593cff1 100644
--- a/cpp-tiny/service/Service.cpp.mustache
+++ b/cpp-tiny/service/Service.cpp.mustache
@@ -21,8 +21,13 @@ void Tiny::Service::begin(std::string url){
 }
 {{/isESP32}}
 
+void Tiny::Service::setAuthorization(const char *username, const char *password)
+{
+    http.setAuthorization(username, password);
+}
+
 void Tiny::Service::addHeader(std::string key, std::string value){
-    http.addHeader(String(key.c_str()), String(value.c_str()));
+    _customheaders.push_back(std::make_tuple(key, value));
 }
 
 void Tiny::Service::addQueryParam(std::string key, std::string value){
@@ -39,6 +44,13 @@ int Tiny::Service::sendRequest(std::string url, const char * type, uint8_t * pay
 
     begin(this->url);
 
+    if(!_customheaders.empty()) {
+        for (auto const& tuple : this->_customheaders) {
+            http.addHeader(String(std::get<0>(tuple).c_str()), String(std::get<1>(tuple).c_str()));
+        }
+    }
+    _customheaders.begin();
+
     int httpCode = http.sendRequest(type, payload, size);
     return httpCode;
 }
diff --git a/cpp-tiny/service/Service.h.mustache b/cpp-tiny/service/Service.h.mustache
index a2f70c1a..7f3a67c7 100644
--- a/cpp-tiny/service/Service.h.mustache
+++ b/cpp-tiny/service/Service.h.mustache
@@ -22,7 +22,8 @@ class Service {
         {{#isESP8266}}
         WiFiClient client;
         {{/isESP8266}}
-        std::string basepath = "https://petstore3.swagger.io/api/v3"; // TODO: change to your url
+        std::string basepath = "{{basePath}}";
+        void setAuthorization(const char *username, const char *password);
         std::string url = "";
         std::string payload = "";
         std::list<std::tuple<std::string, std::string>> queryParams;
@@ -44,6 +45,7 @@ class Service {
         {{/isESP32}}
 
     private:
+        std::list<std::tuple<std::string, std::string>> _customheaders;
         void begin(std::string url);
         void prepareRequest();
         std::string encodeKeyValueTuple(std::list<std::tuple<std::string, std::string>> params);
